Halo2 circuit in mopro (redesigned)

ElusAegis opened the pull request:
Proposition on how we could add Halo2 Circuits to redesigned Mopro.

### Overview
I attempted to be as consistent as possible to `Circom` implementation, resulting in defining two types `Halo2ProveFn` and `Halo2VerifyFn` which we expect the user to implement locally and then use the `key_halo2_circuit_map` as a way to map between different circuits and their implementations of prove and verify functions. 

_Note that as Halo2 is a library, there is **very little** shared components, hence `Mopro` only defines these two types as well as uses `key_halo2_circuit_map` to invoke correct function at run time. _ 

### Halo2 Module in `test-e2e` 
I have also added a `halo2` module in the `test-e2e` crate. It does not contain a halo2 circuit logic (I have extracted it into another repository as discussed before - https://github.com/ElusAegis/halo2-fibonacci-sample.git, which can be forked into Mopro organisation). What it has is logic to interface with the circuit, implementing logic to parse and prove the values. 

As each circuit can use its own types, I did not see a way to abstract it away without enforcing strict user limitations. So `halo2` is an example which users would need to reproduce. (Halo2 requires quite a bit of background knowledge).

### Halo2/Circom Features 

Currently, all users must use both Halo2 and Circom (define `key_halo2_circuit_map` and `zkey_witness_map`.  I have given a thought to how we could be able to disable and enable each of them, however with current design that does not seem to be possible unless we add something like `halo2_placeholder!()` which users need to specify if they do not use Halo2. This does not seem ideal UX.

Hence @vimwitch @vivianjeng , perhaps you have some views on how we could be able to do this? We may need to do something different than the current approach of making a user define a particular function.

ElusAegis commented:
@vimwitch @vivianjeng I have finished the PR. Can you please have a look. 


Key changes were to `app` macros, to enable to have different adapters.

I have it now recursively call `halo2_app` and `circom_app`, which each have a fallback implementation when their respective feature is not implemented.

I have also played with how the `test-e2e/lib.rs` would look like, considering we would need conditionally compile it based on the flags, and I ended up doing it using macros, which are conditionally expanded, thus avoiding to write  `#[cfg(feature = ...)]` before each line.


I have added tests for iOS, however I have not updated Android as I am still unable to compile it.

vimwitch commented on test-e2e/src/halo2/mod.rs:
```diff
@@ -0,0 +1,218 @@
+
+    let public_input = vec![a, b, out];
+
+    create_proof::<KZGCommitmentScheme<Bn256>, ProverGWC<_>, _, _, Blake2bWrite<_, _, _>, _>(
```
I think this logic should also be in `halo2_fibonacci`, specifically the generics about commitment scheme and field.

vimwitch replied:
Ideally the code here is just `halo2_fibonacci::prove(srs_path, proving_key, inputs)`

vimwitch commented on test-e2e/src/lib.rs:
```diff
@@ -1,20 +1,50 @@
 
-app!();
+/// Macro containing the Circom example
+#[cfg(feature = "circom")]
```
I don't think we should do conditional compilation in this crate. test-e2e should test all proving systems.

in the test apps we should eventually add more screens to handle generating different types of proofs

vimwitch replied:
Removing conditional compilation should make this easier to understand. Macros are scary

ElusAegis replied:
Cleaned it up, agree that this is a better way, considering test-e2e should test all proving systems.

vimwitch commented on test-e2e/src/halo2/mod.rs:
```diff
@@ -0,0 +1,218 @@
+use std::collections::HashMap;
```
I don't think `test-e2e` should have more than 1 module (`lib.rs`). All other logic should be delegated to either `mopro-ffi`, or a custom prover (in this case `halo2_fibonacci`).

vimwitch commented on test-e2e/src/halo2/mod.rs:
```diff
@@ -0,0 +1,218 @@
+}
+
+#[cfg(test)]
+mod tests {
```
This isn't testing `mopro-ffi`, these tests should be in `halo2-fibonacci`

vimwitch commented on test-e2e/src/halo2/serialisation.rs:
```diff
@@ -2,28 +2,32 @@ use std::collections::HashMap;
 use std::fmt;
```
also not really `test-e2e`'s concern

vimwitch commented on test-e2e/tests/bindings/test_circom_keccak.swift:
```diff
@@ -73,6 +73,8 @@ do {
 
 } catch let error as MoproError {
   print("MoproError: \(error)")
+  throw error
```
good catch

vimwitch commented on mopro-ffi/src/circom/mod.rs:
```diff
@@ -29,6 +29,75 @@ use color_eyre::Result;
+/// NOTE: YOU CAN ONLY USE THIS MACROS ONCE IN YOUR CODEBASE, IN THE SAME MODULE AS `app!()`
+/// NOTE: TO USE THIS MACRO, YOU MUST HAVE THE `mopro-ffi/circom` FEATURE ENABLED
+#[macro_export]
+macro_rules! set_circom_circuits {
```
Ok so we're using `set_circom_circuits!` to generate a function in the same scope as the above macro. The once-cell above invokes this function to set it's value.

I think we should remove this macro and the once-cell. Users should author the mapping function in rust like before.

Using macros makes it less clear _what_ is happening behind the scenes. It also introduces a non-standard syntax. e.g. even if i know rust i have to figure out this macro.

ElusAegis replied:
Personally (and this is definitely a biased opinion) I though that this would be much more consistent to ask the users to invoke two macros, as opposed to asking to invoking a macro and then implementing a function, which at first does not seem to have clear relevance. 

Especially as rust does not have function overloads, I thought macros was a standard approach. 

As for complexity - macro is a simple solution, however, indeed, it only covers for a function and can be replaced by a function if the user wants.

My take is that if we want a user just use our library and simplify it as much as possible, macro abstracts all of the complexity, while allowing to also take a more complex approach.

I also made sure to add docs to macros (as well as will improve them to mention that a user can do it themselves).

ElusAegis replied:
I will get rid of the `one-cell` as well as the static variable, this was made redundant by the split macro approach.

ElusAegis replied:
I have significantly changed the underlying of the macros, so that we no longer use `one-cell` and instead they generate a match style function, similar to what was originally.

This effectively provides two ways for action for users:
They can simply use a macro (ideal for those who do not know how rust works)
```rust

mopro_ffi::set_circom_circuits! {
    ("multiplier2_final.zkey", multiplier2_witness),
    ("multiplier2_bls_final.zkey", multiplier2bls_witness),
    ("keccak256_256_test_final.zkey", keccak256256test_witness),
}
```rust
fn get_circom_wtns_fn(circuit: &str) -> Result<mopro_ffi::WtnsFn, mopro_ffi::MoproError> {
   match circuit {
      "circuit1.zkey" => Ok(circuit1_witness_fn),
     _ => Err(mopro_ffi::MoproError::CircomError(format!("Unknown ZKEY: {}", circuit).to_string()))
  }
}
```

The doc comment provides a guide to these both options, though I personally think that implementing a function is more complex than using a macro, especially as we already use a macro to initiate `mopro_ffi`.

Furthermore, the function requires error handling, arguably out of scope for the users and only adds future complexity.

vimwitch replied:
I like the change in the rust macro. Providing options as `(str, fn)` tuples is much more clear than just a list.

Having both options is probably the best solution 👍🏻

vimwitch commented on mopro-ffi/src/halo2/mod.rs:
```diff
@@ -1,69 +1,172 @@
+/// NOTE: YOU CAN ONLY USE THIS MACROS ONCE IN YOUR CODEBASE, IN THE SAME MODULE AS `app!()`
+/// NOTE: TO USE THIS MACRO, YOU MUST HAVE THE `mopro-ffi/halo2` FEATURE ENABLED
+#[macro_export]
+macro_rules! set_halo2_proving_circuits {
```
Same here, remove this macro and just let the user write a normal function with a certain name mapping inputs to proof functions.

ElusAegis replied:
Same thoughts as above

vimwitch commented on mopro-ffi/src/halo2/mod.rs:
```diff
@@ -1,69 +1,172 @@
+/// NOTE: YOU CAN ONLY USE THIS MACROS ONCE IN YOUR CODEBASE, IN THE SAME MODULE AS `app!()`
+/// NOTE: TO USE THIS MACRO, YOU MUST HAVE THE `mopro-ffi/halo2` FEATURE ENABLED
+#[macro_export]
+macro_rules! set_halo2_verifying_circuits {
```
Let's try to combine the prove/verify mapping. It should be possible to return a tuple of functions.

ElusAegis replied:
I did combine them under the same macro (the simple approach):
```rust
mopro_ffi::set_halo2_circuits! {
    ("fibonacci_pk", halo2_fibonacci::prove, "fibonacci_vk", halo2_fibonacci::verify)
}
```

However, under the hood they are still different functions, to serve the advanced users who might want to have more tailored logic:

 As effectively a user can select if they want to have a halo2 functionality for only verifying or also proving. And that implies that they should have an option to only specify the circuit they want to be able to verify, while avoiding adding any proving circuits. By merging it all into a single function that removes the flexibility, which will likely be very important for the users.

vimwitch commented on test-e2e/src/lib.rs:
```diff
@@ -1,20 +1,50 @@
-rust_witness::witness!(keccak256256test);
-rust_witness::witness!(hashbenchbls);
+// First, configure the Mopro FFI library
+mopro_ffi::app!();
```
Ok yes, if we remove the `set_` macros we end up with just this as a macro invocation.

vimwitch reviewed (changes requested):
Definitely moving in the right direction. Just need to pull some logic out of `test-e2e` and get rid of a couple macros in favor of written functions.

ElusAegis commented:
@vimwitch thank you for the review, I agree with all of it besides how we show the setup to users. Will get to improving the

ElusAegis commented on test-e2e/src/halo2/mod.rs:
```diff
@@ -0,0 +1,218 @@
+pub(crate) struct FibonacciMoproCircuit;
+
+impl FibonacciMoproCircuit {
+    pub(crate) fn prove(
```
That's a topic I have alluded in the introduction to this PR. Halo2 has a lot of logic unique to how a circuit is launched. And as this is configurable, it is not usually placed into the circuit library. 

We can put it, however users will need to manually do it usually, which means that the sample would require users to likely also explore 'fibonacci' to understand how it works

ElusAegis replied:
I have moved everything to the `halo2_fibonacci`.

ElusAegis commented on .github/workflows/build-and-test.yml:
```diff
@@ -1,190 +1,195 @@
 name: Build and Test
 
 on:
-    push:
```
My bad - my text editor was quite aggressive. Reverted it to the original

ElusAegis commented:
There are further 3 samples which I have created to show how Halo2 would work with Mopro:
	- RSA Full Mopro example  (https://github.com/ElusAegis/halo2-rsa-mopro.git)
        -  Keccak256 Mopro Halo2 Circuit (with use example in mopro/test-e2e) https://github.com/ElusAegis/halo2-keccak-stable.git
        -  Fibonacci Mopro Halo2 Circuit (with use example in mopro/test-e2e) https://github.com/ElusAegis/halo2-fibonacci-sample.git

ElusAegis commented:
## Benchmarking:

**[Keccak256](https://github.com/ElusAegis/halo2-keccak-stable.git):**
1.  _M1 Pro iPhone simulator:_ `116s` to prove and `1.4s` to verify 
2. _iPhone 15 Pro:_ `200s` prove time, `1.1s` verify time

**[RSA](https://github.com/ElusAegis/halo2-rsa-mopro.git):**
1. _M1 Pro iPhone simulator:_` ~450s` proving , `~1s` verifying time (without proving/verifying key generation)
2. _iPhone 15 Pro:_  Crashes due to memory limit exceeded (needs around 5GB RAM, with iPhone limiting to 3GB)

vimwitch commented on test-e2e/Cargo.toml:
```diff
@@ -13,19 +13,33 @@ name = "ios"
+halo2-keccak-256 = { git = "https://github.com/ElusAegis/halo2-keccak-stable.git" }
+
+bincode = "1.3.3"
+halo2curves = { git = "https://github.com/privacy-scaling-explorations/halo2curves", tag = "0.3.2" }
```
Can we get rid of some (most?) of these new deps?

vimwitch commented on test-vectors/halo2/fibonacci_pk:
Can we put a file extension of these? Even just `.bin` to indicate that they are binary files.

ElusAegis replied:
Pushed the changes, hopefully the CI were not broken

vimwitch reviewed (commented):
Good work, this is just about ready to merge

vivianjeng reviewed (approved):
Looks good to me!
Some places I think it can be simplified as I commented
and some comment messages can be written in docs (can be in the following PRs)

ElusAegis commented on .github/workflows/build-and-test.yml:
```diff
@@ -71,7 +61,23 @@ jobs:
+              with:
+                  toolchain: stable
+                  override: true
+            - name: Run e2e halo2 tests
+              # TODO: Fix this custom jar thing
+              run: |
+                  cd test-e2e
+                  curl -L https://repo1.maven.org/maven2/net/java/dev/jna/jna/5.13.0/jna-5.13.0.jar -o jna-5.13.0.jar
+                  CLASSPATH=jna-5.13.0.jar cargo test --test halo2 -- --nocapture
```
Removed the `test-e2e-halo2`, and merged the test task step `test-e2e` of circom , which I have just renamed to test-e2e

ElusAegis commented on mopro-ffi/src/halo2/mod.rs:
```diff
@@ -1,69 +1,163 @@
+            assert!(result.is_ok());
+        } else {
+            panic!("Failed to generate proof")
+        }
```
Done

ElusAegis commented:
@vivianjeng thank you for the review, I have implemented the suggestions. However, after I have merged with the main and updated Cargo.lock it no longer runs the  `iOS-app-simulator` successfully. Yet it does on my local machine with a clean start. 

I do remember we had some problems with non-determinism, could it be one of them? Or was there something in the recent PR that was merged that could have impacted it? I looked at it and nothing seems to be able to break the process at that point.

vimwitch commented:
The iOS simulator on github has intermittent failures 😭. Usually if you read the log it will say something like "timed out" or similar.
